tests/test_terraform.py

```diff
@@ -174,6 +177,32 @@ def test_tf_runner(testdir, tmpdir):
     assert False is tmpdir.join("foo.bar").exists()
 
 
+@pytest.mark.skipif(not shutil.which("terraform"), reason="Terraform binary missing")
+def test_tf_runner_destroy_if_apply_fails(testdir, tmpdir):
+    # ** requires network access to install plugin **
+    with open(tmpdir.join("resources.tf"), "w") as fh:
+        fh.write(
+            """
+resource "local_file" "foo" {
+    content     = "foo!"
+    filename = "${path.module}/foo.bar"
+}
+data "local_file" "foo" {
+    filename = "${path.module}/foo_doesnt_exist.bar"
+    depends_on = [
+        local_file.foo
+    ]
+}
+"""
```

Comment:
Simulate a partial apply:
1 TF apply creates some resources (the local file in this case)
2 TF apply fail on some unrelated resource 

Unfortunately `local_file` resource wont fail if the local file exists already, it just overwrites. So I couldn't use it for the second step.
Instead, we force a failure using a data element to read a non existent file. The side effect of this is the destroy step also fails. But I couldn't find another way to simulate a scenario where a partial apply happens.

Reply:
Ref https://github.com/hashicorp/terraform-provider-local/issues/150